mautrix-{meta,signal,whatsapp}: build with goolm

gmacon opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

After olm gained knownVulnerabilities in #334638, build these bridges using the pure-Go goolm library instead of libolm bindings. I also tried converting mautrix-discord, but it failed to build for what appears to be an upstream reason, but I saw no reason to block the other three. The other mautrix-* bridges in nixpkgs are Python bridges, and I didn't look into building those without olm.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

emilazy commented:
Tagging @sumnerevans here as I expect he’ll have opinions about this – per https://github.com/NixOS/nixpkgs/issues/336052#issuecomment-2298809378 apparently `goolm` has known issues with these bridges.

niklaskorz commented:
I think this should be an option for now, i.e., add it as a package argument so it's easy to override. I'm impartial about which to make the default (libolm has the vulnerability, goolm is potentially not adhering to the spec yet and hasn't been audited).

gmacon commented:
> I think this should be an option for now, i.e., add it as a package argument so it's easy to override.

Do you have any preferences about what shape the option should take? Is it `withGoolm ? false`?

sumnerevans commented:
We (Beeper) tried to use goolm in our next-generation Android client and ran into weird issues. It's not entirely clear whether the issues were due to goolm itself or the way that we were using the library. We ended up just using libolm to reduce the number of variables that could be causing our issues.

I definitely think compilation with goolm should be added as a disabled option for now. Selfishly, it would be great if some people started testing it out more in production and reported any issues with it.

dotlambda commented:
> Selfishly, it would be great if some people started testing it out more in production and reported any issues with it.

I doubt any of those issues would be fixed.
The [upstream repo](https://codeberg.org/DerLukas/goolm) has only seen a total of 16 commits by a single author.
There's also nothing that indicates it's more secure than libolm. Rather the opposite.

emilazy commented:
Mautrix uses [their own fork](https://github.com/mautrix/go/tree/f56905a27645bb38b2da452fc409c2a910d64dfc/crypto/goolm).

Commit:
mautrix-{meta,signal,whatsapp}: Optionally build against goolm

After olm gained knownVulnerabilities in #334638, allow building these
bridges using the pure-Go goolm library instead of libolm bindings.

gmacon commented:
Ok, here we go. OfBorg and Hydra won't build these as-is, but I tested the builds with commands like

```
nix build --impure --expr 'let p = import ./. {}; in p.mautrix-whatsapp.override { withGoolm = true; }'
```

I intend to use these options for my self-hosted Beeper bridges, and I'll report any issues I run into.

niklaskorz reviewed (approved):
Tested the whatsapp and signal bridges, both seem to work just fine with goolm enabled

Edit: looks like replies on the whatsapp bridge are broken with goolm
Edit 2: same on the Signal bridge, all incoming messages are broken with goolm. Nonetheless I'm not opposed to merging this as it's opt-in right now.

Ma27 reviewed (changes requested):
Given the current issues with this flag we should add a warning as comment above. Given that these flags aren't discoverable at all, you'll see the warning when you learn about the flag (by reading the expression).

gmacon commented:
How about:

> This option enables the use of the goolm library instead of libolm for end-to-end encryption. Using goolm is not officially supported by the mautrix developers, but they are interested in people trying it out and reporting any issues they run into.

@sumnerevans , I'm particularly interested in your feedback here to make sure that I don't misrepresent what Beeper wants to promise/request.

sumnerevans commented:
We marked it as experimental as of https://github.com/mautrix/go/releases/tag/v0.17.0, so maybe:

> This option enables the use of an experimental pure-Go implementation of the Olm protocol instead of libolm for end-to-end encryption. Using goolm is not recommended by the mautrix developers, but they are interested in people trying it out in non-production-critical environments and reporting any issues they run into.

Given the issues that people have already mentioned, I think we should warn people about using this in production-critical environments.

Ma27 reviewed (approved):
Approach is OK for me.
Haven't tried the goolm variant out so far, but @sumnerevans having approved this is OK change-wise, so let's merge.

Henry-Hiles commented:
Is the discord goolm still broken?

Henry-Hiles commented:
If not, I'll make a PR to add an option for discord too.

Henry-Hiles commented:
Appears its still broken.
